Reuse netty buffers during backup

lutovich opened the pull request:
During backup, bytes from store files are transferred using netty `ChannelBuffer`s. Such buffers were created for each chunk of data that was send out. Default chunk size is 4M so for big stores amount of created buffers was huge.

This PR introduces a chunking channel buffer that is able to reuse `ChannelBuffer`s. Currently it is used only for backup because excessive buffer creation was only noticed there.

`BufferReusingChunkingChannelBuffer` uses a queue of free buffers and subscribes to 'write completed' notifications with a listener that clears the used buffer and puts it on the queue of free buffers. Essentially this chunking channel buffer trades allocation of dynamic netty buffers for allocation of `ChannelFutureListener`s. This is a right thing to do because `ChannelBuffer` is an array based structure while `ChannelFutureListener` is only an anonymous class that captures a single `ChannelBuffer`.

This change was tested on a 2.5GB store with full backups running in a tight loop. 1 minute JFR recording showed following results:
- without buffer reuse: 14592 byte[] instances, total size 26.05GB
- with buffer reuse: 2436 byte[] instances, total size 9.13GB

So improvement in object allocation would probably be only visible for big stores and full backups.

lutovich committed:
Reuse netty buffers during backup

During backup, bytes from store files are transferred using netty
ChannelBuffers. Such buffers were created for each chunk of data that
was send out. Default chunk size is 4M so for big stores amount of
created buffers was huge.

This commit introduces a chunking channel buffer that is able to reuse
ChannelBuffers. Currently it is used only for backup because excessive
buffer creation was only noticed there.

BufferReusingChunkingChannelBuffer uses a queue of free buffers and
subscribes to 'write completed' notifications with a listener that
clears the used buffer and puts it on the queue of free buffers.
Essentially this chunking channel buffer trades allocation of dynamic
netty buffers for allocation of ChannelFutureListeners. This is a right
thing to do because ChannelBuffer is an array based structure while
ChannelFutureListener is only an anonymous class that captures a
single ChannelBuffer.

This change was tested on a 2.5GB store with full backups running in a
tight loop. 1 minute JFR recording showed following results:
 * without buffer reuse: 14592 byte[] instances, total size 26.05GB
 * with buffer reuse: 2436 byte[] instances, total size 9.13GB

So improvement in object allocation would probably be only visible for
big stores and full backups.

thobe commented:
Looks pretty good. I wonder what evidence we have that `5` is the appropriate number for `MAX_WRITE_AHEAD_CHUNKS`?

The only thing that makes me uneasy is all the mocking in the testing. I really don't like mocks, and the way it is done here, with spies, it makes the test extremely tightly coupled to the code, plus the code had to grow some weird appendages in order to support the testing, which is also undesirable. I wonder if we could structure the code differently to make it easier to test without resorting to mocks, but I don't think it is important enough to hold this change back.

lutovich commented:
I do not actually know why `MAX_WRITE_AHEAD_CHUNKS = 5`. This is something internal to `ChunkingChannelBuffer` which has no tests. Motivation for using it as queue size is that it is max number of `ChunnelBuffer`s that writer (`ChunkingChannelBuffer`) could be ahead of the reader (`DechunkingChannelBuffer`). Writer stops writing when `>= MAX_WRITE_AHEAD_CHUNKS` were written: https://github.com/neo4j/neo4j/blob/2.3/enterprise/com/src/main/java/org/neo4j/com/ChunkingChannelBuffer.java#L555.

Testing could be done with less mocking if we pass in a factory of `ChannelBuffer`s to `BufferReusingChunkingChannelBuffer`. Than `spy`s would not be needed. Does this sound sensible?

thobe commented:
@lutovich that sounds very sensible. Could you make that as a separate commit first, so that I can look at it in isolation? (and then possibly squash it later)

Thanks for the info on `MAX_WRITE_AHEAD_CHUNKS`, sound sensible. Have you done any tracking of when / how often we allocate buffers that fail to go back to the pool? Because it is unclear to me if that relationship between readers and writers is the only thing that uses buffers.

lutovich committed:
Improved testability of BufferReusingChunkingChannelBuffer

Now it accepts a factory of ChannelBuffers which makes testing more pleasant
compared to method overriding.
Rewritten corresponding test to use this new ability.

lutovich commented:
@thobe Improved testability of `BufferReusingChunkingChannelBuffer` a bit.
Made a quick test for buffer reuse with this code https://gist.github.com/lutovich/8d1c36adb8e72ca672a4 which simply starts HA database on a 2.5GB store and makes 5 online full backups in a row. `BufferReusingChunkingChannelBuffer#newChannelBuffer()` increments a counter when it has to create a new buffer. Here are results:

```
1. unchanged code - 6 buffers created
2. Thread.sleep(1) in DechunkingChannelBuffer#readNextChunk() - 5 buffers created
3. Thread.sleep(10) in DechunkingChannelBuffer#readNextChunk() - 9 buffers created
4. Thread.sleep(100) in DechunkingChannelBuffer#readNextChunk() - 25 buffers created
5. Thread.sleep(500) in DechunkingChannelBuffer#readNextChunk() - 25 buffers created
```

Sleeping was meant to emulate slow receiver.
Looks like `ChunkingChannelBuffer#waitForClientToCatchUpOnReadingChunks()` makes backup use no more than `MAX_WRITE_AHEAD_CHUNKS` buffers.

thobe commented:
Very nice.

thobe commented:
I have verified that this does indeed reduce allocation significantly, as well as reduce GC load.
